### PR TITLE
DW-2220.

### PR DESCRIPTION
**Pull request by DanielChicot**
Integration test for oversized cipher text payload.

Integration test for missing or malformed parameters.

Removed initialisation vector from responses and documentation.

Additional unit tests to fully exercise providers.

### CODE REVIEW

File: src/main/java/uk/gov/dwp/dataworks/controller/DataKeyController.java [hunk 1]
```diff
@@ -4,21 +4,24 @@
...
 
 @SuppressWarnings("unused")
 @RestController
 @RequestMapping("/datakey")
 @Api(value="datakey")
 public class DataKeyController {
 
-    private DataKeyService dataKeyService;
+    private final DataKeyService dataKeyService;
+    private final static Logger LOGGER = LoggerFactory.getLogger(DataKeyController.class);
```
Comment (nicholasjgreen) [inline src/main/java/uk/gov/dwp/dataworks/controller/DataKeyController.java]: This logger does not seem to be in use

**Comment by nicholasjgreen** [comment]
I probably should have mentioned this in the original ticket (my bad) but it is also possible to configure springboot to restrict content sizes https://stackoverflow.com/questions/33232849/increase-http-post-maxpostsize-in-spring-boot/41467782

File: src/main/java/uk/gov/dwp/dataworks/errors/DataKeyDecryptionException.java [hunk 2]
```diff
@@ -5,7 +5,7 @@
 
 @ResponseStatus(HttpStatus.SERVICE_UNAVAILABLE)
 public class DataKeyDecryptionException extends RuntimeException {
-    public DataKeyDecryptionException() {
-        super("Failed to decrypt this data key due to an internal error. Try again later.");
+    public DataKeyDecryptionException(Throwable ex) {
+        super("Failed to decrypt this data key due to an internal error. Try again later.", ex);
```
[the remark this answers is not in the document]
Comment (DanielChicot) [reply src/main/java/uk/gov/dwp/dataworks/errors/DataKeyDecryptionException.java]: Removed cause arguments from all exception whose details are potentially sent over the wire. Replaced with fairly anodyne error message.

File: src/main/java/uk/gov/dwp/dataworks/provider/KMSCurrentKeyIdProvider.java [hunk 3]
```diff
@@ -1,33 +1,34 @@
...
                     .withWithDecryption(false);
-            GetParameterResult result = client.getParameter(request);
+            GetParameterResult result = awsSimpleSystemsManagementClient.getParameter(request);
             return result.getParameter().getValue();
         }
-        catch(Exception ex) {
-            logger.error("Exception caught while communicating with parameter store", ex);
-            throw new CurrentKeyIdException();
+        catch (Exception e) {
+            throw new CurrentKeyIdException(e);
```
[the remark this answers is not in the document]
Comment (DanielChicot) [reply src/main/java/uk/gov/dwp/dataworks/provider/KMSCurrentKeyIdProvider.java]: Reintroduced log.error messages that were issued prior to raising exception sent over the wire. Included stack trace.

File: src/main/java/uk/gov/dwp/dataworks/controller/DataKeyController.java (hunk 1 above)
Comment (DanielChicot) [reply src/main/java/uk/gov/dwp/dataworks/controller/DataKeyController.java]: Removed unused Logger from DatakeyController

**Comment by DanielChicot** [comment]
I have raised DW-2369 in respect of the max upload size, with findings.

Also actioned Mark's feedback in which he requested that exception error messages are validated.

File: src/test/java/uk/gov/dwp/dataworks/provider/KMSDataKeyGeneratorProviderTest.java [hunk 4]
```diff
@@ -1,24 +1,157 @@
...
+    @Test
+    public void generateDataKey() {
+        String dataKeyEncryptionKeyId = "DATAKEYENCRYPTIONKEYID";
+        String plainTextKey = "PLAINTEXTKEY";
+        String encryptedDataKey = "ENCRYPTEDDATAKEY";
+        GenerateDataKeyResult result = new GenerateDataKeyResult();
+        result.setKeyId(dataKeyEncryptionKeyId);
+        result.setPlaintext(ByteBuffer.wrap(plainTextKey.getBytes()));
+        result.setCiphertextBlob(ByteBuffer.wrap(encryptedDataKey.getBytes()));
+        GenerateDataKeyRequest generateDataKeyRequest = new GenerateDataKeyRequest();
```
Comment (nicholasjgreen) [inline src/test/java/uk/gov/dwp/dataworks/provider/KMSDataKeyGeneratorProviderTest.java]: Feels a bit words to need to use setters on a DTO. Are you sure you don't want to set them in the constructor? The request is not valid without both being set

**Comment by nicholasjgreen** [comment]
@DanielChicot looking pretty good. I like that you're exercising the exception cases. I've written one thing, but TBH its pretty minor and you could argue that I'm wrong

**Comment by DanielChicot** [comment]
GenerateDataKeyResult is an AWS class and only has a no-arg constructor so I haven't been able to implement the final recommendation.

**Pull request merged**